src/main/java/uk/gov/hmcts/cmc/claimstore/services/ccd/callbacks/generalletter/GeneralLetterService.java

```diff
@@ -0,0 +1,160 @@
+                authorisation, templateId);
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .data(ImmutableMap.of(
+                    letterType,
+                    CCDDocument.builder().documentUrl(docAssemblyResponse.getRenditionOutputLocation())))
+                .build();
+        } catch (Exception e) {
+            logger.info("General Letter creating and preview failed", e);
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .errors(Collections.singletonList(ERROR_MESSAGE))
+                .build();
+        }
+    }
+
+    public CallbackResponse printAndUpdateCaseDocuments(CaseDetails caseDetails, String authorisation) {
+        try {
+            CCDCase ccdCase = caseDetailsConverter.extractCCDCase(caseDetails);
+            CCDDocument draftLetterDoc = ccdCase.getDraftLetterDoc();
+            Claim claim = caseDetailsConverter.extractClaim(caseDetails);
+            printLetter(authorisation, draftLetterDoc, claim);
+            CCDCase updatedCase = ccdCase.toBuilder()
+                .caseDocuments(updateCaseDocumentsWithOrder(ccdCase, draftLetterDoc))
+                .build();
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .data(caseDetailsConverter.convertToMap(updatedCase))
+                .build();
+
+        } catch (Exception e) {
+            logger.info("General Letter printing and case documents update failed", e);
+            return AboutToStartOrSubmitCallbackResponse
+                .builder()
+                .errors(Collections.singletonList(ERROR_MESSAGE))
+                .build();
+        }
+    }
+
+    private List<CCDCollectionElement<CCDClaimDocument>> updateCaseDocumentsWithOrder(
+        CCDCase ccdCase,
+        CCDDocument draftLetterDoc
+    ) {
+        CCDCollectionElement<CCDClaimDocument> claimDocument = CCDCollectionElement.<CCDClaimDocument>builder()
+            .value(CCDClaimDocument.builder()
+                .documentLink(draftLetterDoc)
+                .documentName(getDocumentNumber(ccdCase))
+                .createdDatetime(LocalDateTime.now(clock.withZone(UTC_ZONE)))
+                .documentType(GENERAL_LETTER)
+                .build())
+            .build();
+        return ImmutableList.<CCDCollectionElement<CCDClaimDocument>>builder()
+            .addAll(ccdCase.getCaseDocuments())
+            .add(claimDocument)
+            .build();
+    }
+
+    private String getDocumentNumber(CCDCase ccdCase) {
+        String number = String.valueOf((ccdCase.getCaseDocuments()
+            .stream().filter(c ->
+                c.getValue().getDocumentType()
+                    .equals(GENERAL_LETTER) && c.getValue().getDocumentName()
+                    .contains(LocalDate.now().toString())
+            ).collect(Collectors.toList()).size() + 1));
```

Comment:
Try to keep stream operations as simple and atomic as possible:
```suggestion
            .stream()
            .filter(c -> c.getValue().getDocumentType().equals(GENERAL_LETTER))
            .filter(c -> c.getValue().getDocumentName().contains(LocalDate.now().toString()))
            .collect(Collectors.toList()).size() + 1));
```